R/binary_beta.R

```diff
@@ -0,0 +1,2567 @@
+    coords[, seq_len(k_use), drop = FALSE]
+  } else {
+    matrix(
+      numeric(0L),
+      nrow = n,
+      ncol = 0L,
+      dimnames = list(labels, character(0L))
+    )
+  }
+
+  sample_coords <- tibble::as_tibble(coords_k, rownames = "sample_id")
+
+  # ----------------------------------------------------------------------------
+  # 4) variance explained (based on positive eigenvalues)
+  # ----------------------------------------------------------------------------
+  .ph_log_info("summarizing eigenvalues and variance explained.")
+  pos_eig <- pmax(eig_vals, 0)
+  sum_pos <- sum(pos_eig, na.rm = TRUE)
+  neg_eig <- pmin(eig_vals, 0)
+  sum_neg <- sum(abs(neg_eig), na.rm = TRUE)
+
+  if (sum_pos > 0 && k_use > 0L) {
+    pct_axes <- 100 * pos_eig[seq_len(k_use)] / sum_pos
+    pct_other <- if (length(pos_eig) > k_use) {
+      100 * sum(pos_eig[(k_use + 1L):length(pos_eig)], na.rm = TRUE) / sum_pos
+    } else {
+      0
+    }
+  } else {
+    pct_axes <- rep(NA_real_, k_use)
+    pct_other <- NA_real_
+  }
+
+  names(pct_axes) <- paste0("%PCoA", seq_len(k_use))
+  var_explained <- tibble::as_tibble_row(
+    c(as.list(round(pct_axes, 3)), `%Other` = round(pct_other, 3))
+  )
+
+  min_eig <- if (length(eig_vals) > 0L) min(eig_vals, na.rm = TRUE) else NA_real_
+  n_neg <- sum(eig_vals < 0, na.rm = TRUE)
+  frac_neg <- if (length(eig_vals) > 0L) n_neg / length(eig_vals) else NA_real_
+  ratio_neg_pos <- if (sum_pos > 0) sum_neg / sum_pos else NA_real_
+
+  eigen_diagnostics <- tibble::as_tibble_row(list(
+    sum_negative = sum_neg,
+    sum_positive = sum_pos,
+    ratio_negative_positive = ratio_neg_pos,
+    min_eigenvalue = min_eig,
+    n_negative = n_neg,
+    frac_negative = frac_neg
+  ))
+
+  correction_infos <- tibble::as_tibble_row(list(
+    corrected = is_corrected,
+    correction_method = neg_correction,
+    correction_add = correction_add,
+    correction_note = correction_note
+  ))
+
+  # ----------------------------------------------------------------------------
+  # 5) feature associations (requires abundances attribute)
+  # ----------------------------------------------------------------------------
+  .ph_log_info(
+    paste0("computing feature associations: ", feature_assoc)
+  )
+  feature_associations <- tibble::tibble()
+
+  X <- attr(dist_obj, "abundances")
+  if (is.null(X)) {
+    .ph_warn(
+      "no 'abundances' attribute found on `dist_obj`;
+      skipping feature associations."
+    )
+  } else if (identical(feature_assoc, "none")) {
+    feature_associations <- tibble::tibble()
+  } else if (k_use < 1L) {
+    # no axes, nothing to load
+    feature_associations <- tibble::tibble()
+  } else {
+    X <- as.matrix(X)
+
+    coords_ids <- rownames(coords)
+    X_ids <- rownames(X)
+
+    if (is.null(coords_ids) || is.null(X_ids)) {
+      .ph_warn(
+        "row names missing in coordinates or 'abundances'; cannot align samples
+        for feature associations."
+      )
+    } else {
```

Comment:
You have an anti-pattern here where you have a "brackets pyramid".
Once this is in a helper function, it will be very easy to just early `return` an empty result on every such problem instead of stacking the if clauses on top of one another.